src/core/integrators/brownian_inline.hpp

```diff
@@ -0,0 +1,570 @@
+/*
+ * Copyright (C) 2010-2019 The ESPResSo project
+ * Copyright (C) 2002,2003,2004,2005,2006,2007,2008,2009,2010
+ *   Max-Planck-Institute for Polymer Research, Theory Group
+ *
+ * This file is part of ESPResSo.
+ *
+ * ESPResSo is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * ESPResSo is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+/** \file */
+
+#ifndef BROWNIAN_INLINE_HPP
+#define BROWNIAN_INLINE_HPP
+
+#include "random.hpp"
+#include "thermostat.hpp"
+
+extern BrownianThermostat brownian;
+
+/** Propagate position: viscous drag driven by conservative forces.
+ *  From eq. (14.39) in @cite Schlick2010.
+ *  @param[in,out] p              Particle
+ *  @param[in]     dt             Time interval
+ */
+inline void bd_drag(Particle &p, double dt) {
+  // The friction tensor Z from the Eq. (14.31) of Schlick2010:
+  Thermostat::GammaType local_gamma;
+
+#ifdef BROWNIAN_PER_PARTICLE
+  if (p.p.gamma >= Thermostat::GammaType{}) {
+    local_gamma = p.p.gamma;
+  } else
+#endif
+  {
+    local_gamma = brownian.gamma;
+  }
+
+  bool aniso_flag; // particle anisotropy flag
+
+#ifdef PARTICLE_ANISOTROPY
+  // Particle frictional isotropy check.
+  aniso_flag =
+      (local_gamma[0] != local_gamma[1]) || (local_gamma[1] != local_gamma[2]);
+#endif
+
+  Utils::Vector3d force_body;
+  Utils::Vector3d delta_pos_body, delta_pos_lab;
+
+#ifdef PARTICLE_ANISOTROPY
+  if (aniso_flag) {
+    force_body = convert_vector_space_to_body(p, p.f.f);
+  }
+#endif
+
+  for (int j = 0; j < 3; j++) {
+    // Second (deterministic) term of the Eq. (14.39) of Schlick2010.
+    // Only a conservative part of the force is used here
+#ifdef PARTICLE_ANISOTROPY
+    if (aniso_flag) {
+      delta_pos_body[j] = force_body[j] * dt / (local_gamma[j]);
+    } else {
+#ifdef EXTERNAL_FORCES
+      if (!(p.p.ext_flag & COORD_FIXED(j)))
+#endif
+      {
+        p.r.p[j] += p.f.f[j] * dt / (local_gamma[j]);
```

Comment:
still missing a check for virtual particles here and probably at other places too